### PR TITLE
CBL-6009: Provide an option to enable full sync in the database

### PR DESCRIPTION
**Pull request by jianminzhao**
Added a flag to C4DatabaseFlags, kC4DB_DiskSyncFull. This flag is passed to DataFile::Options, which is used as we request a connection to the SQLite database.

### CODE REVIEW

**Review by bmeike — COMMENTED** [review]
:white_check_mark:

File: C/include/c4DatabaseTypes.h [hunk 1]
```diff
@@ -37,6 +37,7 @@ typedef C4_OPTIONS(uint32_t, C4DatabaseFlags) {
     kC4DB_VersionVectors= 0x08, ///< Upgrade DB to version vectors instead of rev trees [EXPERIMENTAL]
     kC4DB_NoUpgrade     = 0x20, ///< Disable upgrading an older-version database
     kC4DB_NonObservable = 0x40, ///< Disable database/collection observers, for slightly faster writes
+    kC4DB_DiskSyncFull  = 0x80, ///< Write to disk after each transaction
```
Comment (pasin) [inline C/include/c4DatabaseTypes.h:40]: Maybe only SyncFull or FullSync?

**Review by pasin — APPROVED** [review]

File: C/include/c4DatabaseTypes.h (hunk 1 above)
Comment (borrrden) [reply C/include/c4DatabaseTypes.h:40]: The comment is also probably misleading.  All transactions write to the disk after they finish successfully, but the full sync also introduces a manual flush.

File: LiteCore/tests/c4BaseTest.cc [hunk 2]
```diff
@@ -144,6 +146,30 @@ TEST_CASE("C4Error Reporting Macros", "[Errors][C]") {
 }
 
 
+TEST_CASE_METHOD(C4Test, "Database Flag FullSync", "[Database][C]") {
```
Comment (borrrden) [inline LiteCore/tests/c4BaseTest.cc:149]: There should probably at least be a manual inspection of one of these databases to confirm that the full sync is taking effect at the SQLite setting.

File: C/include/c4DatabaseTypes.h (hunk 1 above)
Comment (jianminzhao) [reply C/include/c4DatabaseTypes.h:40]: That was my choice in the first attempt, but afterwards, I feel "disk" makes it clearer. It is about the sync between the disk and memory.

File: LiteCore/tests/c4BaseTest.cc (hunk 2 above)
Comment (jianminzhao) [reply LiteCore/tests/c4BaseTest.cc:149]: I did manual inspection here at 
`            _exec(format("PRAGMA cache_size=%d; "            // Memory cache
                         "PRAGMA mmap_size=%d; "             // Memory-mapped reads
                         "PRAGMA synchronous=%s; "           // Speeds up commits
                         "PRAGMA journal_size_limit=%lld; "  // Limit WAL disk usage
                         "PRAGMA case_sensitive_like=true; "   // Case sensitive LIKE, for N1QL compat
                         "PRAGMA fullfsync=ON",              // Attempt to mitigate damage due to sudden loss of power (iOS / macOS)
                         -(int)kCacheSize/1024, kMMapSize, getOptions().diskSyncFull?"full":"normal",
                         (long long)kJournalSize));
`
by breaking there and inspected the string in either case constructed by "format", particularly, "full" vs "normal".

Comment (jianminzhao) [reply LiteCore/tests/c4BaseTest.cc:149]: Is it sufficient?

File: C/include/c4DatabaseTypes.h (hunk 1 above)
Comment (jianminzhao) [reply C/include/c4DatabaseTypes.h:40]: "Flush to disk after each transaction" ?
@borrrden

File: LiteCore/tests/c4BaseTest.cc (hunk 2 above)
Comment (borrrden) [reply LiteCore/tests/c4BaseTest.cc:149]: That seems reasonable

**Review by borrrden — APPROVED** [review]

**Pull request merged**